### ISSUE TITLE
Kafka cannot reach zookeeper. Endpoints not correctly set? 

### ISSUE THREAD

**Issue by BenCoughlan15** [strimzi/strimzi-kafka-operator#1264]
Hi again!

Right, so I've got my zk up and happily chugging along, but it seems like kafka can't connect. 

It also oddly seems to be pointing at localhost:2181 to find the ZK nodes, which seems incorrect. 

Any ideas? I'm happy to provide descriptions, etc. 

Cheers,
Ben

```
KAFKA_BROKER_ID=1
KAFKA_LOG_DIRS=/var/lib/kafka/kafka-log1
Preparing truststore for replication listener
Adding /opt/kafka/cluster-ca-certs/ca.crt to truststore /tmp/kafka/cluster.truststore.p12 with alias ca
Certificate was added to keystore
Preparing truststore for replication listener is complete
Preparing keystore for replication and clienttls listener
Preparing keystore for replication and clienttls listener is complete
Preparing truststore for clienttls listener
Adding /opt/kafka/client-ca-certs/ca.crt to truststore /tmp/kafka/clients.truststore.p12 with alias ca
Certificate was added to keystore
Preparing truststore for clienttls listener is complete
Starting Kafka with configuration:
broker.id=1
broker.rack=

# Listeners
listeners=REPLICATION://0.0.0.0:9091,EXTERNAL://0.0.0.0:9094
advertised.listeners=REPLICATION://kafka-kafka-1.kafka-kafka-brokers.kafka.svc.cluster.local:9091,EXTERNAL://ip-10-0-21-6.eu-west-2.compute.internal:30625
listener.security.protocol.map=REPLICATION:SSL,EXTERNAL:SSL
inter.broker.listener.name=REPLICATION

# Zookeeper
zookeeper.connect=localhost:2181
zookeeper.connection.timeout.ms=6000

# Logs
log.dirs=/var/lib/kafka/kafka-log1

# TLS / SSL
ssl.keystore.password=mZi4eq-SYmM7vnWZgnRn6z8asEGfuECH
ssl.truststore.password=mZi4eq-SYmM7vnWZgnRn6z8asEGfuECH
ssl.keystore.type=PKCS12
ssl.truststore.type=PKCS12
ssl.endpoint.identification.algorithm=HTTPS
ssl.secure.random.implementation=SHA1PRNG

listener.name.replication.ssl.keystore.location=/tmp/kafka/cluster.keystore.p12
listener.name.replication.ssl.truststore.location=/tmp/kafka/cluster.truststore.p12
listener.name.replication.ssl.client.auth=required

sasl.enabled.mechanisms=



# EXTERNAL interface configuration
listener.name.external.ssl.keystore.location=/tmp/kafka/cluster.keystore.p12
listener.name.external.ssl.truststore.location=/tmp/kafka/clients.truststore.p12
# EXTERNAL listener authentication
listener.name.external.ssl.client.auth=none

# Authorization configuration
authorizer.class.name=


# Provided configuration
transaction.state.log.replication.factor=3
offsets.topic.replication.factor=3
transaction.state.log.min.isr=2


OpenJDK 64-Bit Server VM warning: If the number of processors is expected to increase from one, then you should configure the number of parallel GC threads appropriately using -XX:ParallelGCThreads=N
OpenJDK 64-Bit Server VM warning: Cannot open file /opt/kafka/kafkaServer-gc.log due to Permission denied

2019-01-28 14:33:18,573 INFO Registered kafka:type=kafka.Log4jController MBean (kafka.utils.Log4jControllerRegistration$) [main]
2019-01-28 14:33:19,014 INFO starting (kafka.server.KafkaServer) [main]
2019-01-28 14:33:19,015 INFO Connecting to zookeeper on localhost:2181 (kafka.server.KafkaServer) [main]
2019-01-28 14:33:19,032 INFO [ZooKeeperClient] Initializing a new session to localhost:2181. (kafka.zookeeper.ZooKeeperClient) [main]
2019-01-28 14:33:19,036 INFO Client environment:zookeeper.version=3.4.13-2d71af4dbe22557fda74f9a9b4309b15a7487f03, built on 06/29/2018 00:39 GMT (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:host.name=kafka-kafka-1.kafka-kafka-brokers.kafka.svc.cluster.local (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:java.version=1.8.0_191 (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:java.vendor=Oracle Corporation (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:java.home=/usr/lib/jvm/java-1.8.0-openjdk-1.8.0.191.b12-1.el7_6.x86_64/jre (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:java.class.path=/opt/kafka/bin/../libs/activation-1.1.1.jar:/opt/kafka/bin/../libs/aopalliance-repackaged-2.5.0-b42.jar:/opt/kafka/bin/../libs/argparse4j-0.7.0.jar:/opt/kafka/bin/../libs/audience-annotations-0.5.0.jar:/opt/kafka/bin/../libs/commons-lang3-3.5.jar:/opt/kafka/bin/../libs/connect-api-2.0.0.jar:/opt/kafka/bin/../libs/connect-basic-auth-extension-2.0.0.jar:/opt/kafka/bin/../libs/connect-file-2.0.0.jar:/opt/kafka/bin/../libs/connect-json-2.0.0.jar:/opt/kafka/bin/../libs/connect-runtime-2.0.0.jar:/opt/kafka/bin/../libs/connect-transforms-2.0.0.jar:/opt/kafka/bin/../libs/guava-20.0.jar:/opt/kafka/bin/../libs/hk2-api-2.5.0-b42.jar:/opt/kafka/bin/../libs/hk2-locator-2.5.0-b42.jar:/opt/kafka/bin/../libs/hk2-utils-2.5.0-b42.jar:/opt/kafka/bin/../libs/jackson-annotations-2.9.6.jar:/opt/kafka/bin/../libs/jackson-core-2.9.6.jar:/opt/kafka/bin/../libs/jackson-databind-2.9.6.jar:/opt/kafka/bin/../libs/jackson-jaxrs-base-2.9.6.jar:/opt/kafka/bin/../libs/jackson-jaxrs-json-provider-2.9.6.jar:/opt/kafka/bin/../libs/jackson-module-jaxb-annotations-2.9.6.jar:/opt/kafka/bin/../libs/javassist-3.22.0-CR2.jar:/opt/kafka/bin/../libs/javax.annotation-api-1.2.jar:/opt/kafka/bin/../libs/javax.inject-1.jar:/opt/kafka/bin/../libs/javax.inject-2.5.0-b42.jar:/opt/kafka/bin/../libs/javax.servlet-api-3.1.0.jar:/opt/kafka/bin/../libs/javax.ws.rs-api-2.1.jar:/opt/kafka/bin/../libs/jaxb-api-2.3.0.jar:/opt/kafka/bin/../libs/jersey-client-2.27.jar:/opt/kafka/bin/../libs/jersey-common-2.27.jar:/opt/kafka/bin/../libs/jersey-container-servlet-2.27.jar:/opt/kafka/bin/../libs/jersey-container-servlet-core-2.27.jar:/opt/kafka/bin/../libs/jersey-hk2-2.27.jar:/opt/kafka/bin/../libs/jersey-media-jaxb-2.27.jar:/opt/kafka/bin/../libs/jersey-server-2.27.jar:/opt/kafka/bin/../libs/jetty-client-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-continuation-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-http-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-io-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-security-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-server-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-servlet-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-servlets-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jetty-util-9.4.11.v20180605.jar:/opt/kafka/bin/../libs/jopt-simple-5.0.4.jar:/opt/kafka/bin/../libs/kafka-clients-2.0.0.jar:/opt/kafka/bin/../libs/kafka-log4j-appender-2.0.0.jar:/opt/kafka/bin/../libs/kafka-streams-2.0.0.jar:/opt/kafka/bin/../libs/kafka-streams-examples-2.0.0.jar:/opt/kafka/bin/../libs/kafka-streams-scala_2.12-2.0.0.jar:/opt/kafka/bin/../libs/kafka-streams-test-utils-2.0.0.jar:/opt/kafka/bin/../libs/kafka-tools-2.0.0.jar:/opt/kafka/bin/../libs/kafka_2.12-2.0.0-sources.jar:/opt/kafka/bin/../libs/kafka_2.12-2.0.0.jar:/opt/kafka/bin/../libs/log4j-1.2.17.jar:/opt/kafka/bin/../libs/lz4-java-1.4.1.jar:/opt/kafka/bin/../libs/maven-artifact-3.5.3.jar:/opt/kafka/bin/../libs/metrics-core-2.2.0.jar:/opt/kafka/bin/../libs/osgi-resource-locator-1.0.1.jar:/opt/kafka/bin/../libs/plexus-utils-3.1.0.jar:/opt/kafka/bin/../libs/reflections-0.9.11.jar:/opt/kafka/bin/../libs/rocksdbjni-5.7.3.jar:/opt/kafka/bin/../libs/scala-library-2.12.6.jar:/opt/kafka/bin/../libs/scala-logging_2.12-3.9.0.jar:/opt/kafka/bin/../libs/scala-reflect-2.12.6.jar:/opt/kafka/bin/../libs/slf4j-api-1.7.25.jar:/opt/kafka/bin/../libs/slf4j-log4j12-1.7.25.jar:/opt/kafka/bin/../libs/snappy-java-1.1.7.1.jar:/opt/kafka/bin/../libs/validation-api-1.1.0.Final.jar:/opt/kafka/bin/../libs/zkclient-0.10.jar:/opt/kafka/bin/../libs/zookeeper-3.4.13.jar:/opt/prometheus/jmx_prometheus_javaagent.jar (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:java.library.path=/usr/java/packages/lib/amd64:/usr/lib64:/lib64:/lib:/usr/lib (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:java.io.tmpdir=/tmp (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:java.compiler=<NA> (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:os.name=Linux (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,036 INFO Client environment:os.arch=amd64 (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,037 INFO Client environment:os.version=3.10.0-862.11.6.el7.x86_64 (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,037 INFO Client environment:user.name=kafka (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,037 INFO Client environment:user.home=/home/kafka (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,037 INFO Client environment:user.dir=/opt/kafka (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,037 INFO Initiating client connection, connectString=localhost:2181 sessionTimeout=6000 watcher=kafka.zookeeper.ZooKeeperClient$ZooKeeperClientWatcher$@34a3d150 (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:19,045 INFO [ZooKeeperClient] Waiting until connected. (kafka.zookeeper.ZooKeeperClient) [main]
2019-01-28 14:33:19,046 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:19,050 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:20,151 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:20,152 INFO Socket connection established to localhost/127.0.0.1:2181, initiating session (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:20,157 WARN Session 0x0 for server localhost/127.0.0.1:2181, unexpected error, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:192)
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
	at org.apache.zookeeper.ClientCnxnSocketNIO.doIO(ClientCnxnSocketNIO.java:68)
	at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:366)
	at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1145)
2019-01-28 14:33:22,177 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:22,178 INFO Socket connection established to localhost/127.0.0.1:2181, initiating session (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:22,184 WARN Session 0x0 for server localhost/127.0.0.1:2181, unexpected error, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
	at sun.nio.ch.IOUtil.read(IOUtil.java:192)
	at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:380)
	at org.apache.zookeeper.ClientCnxnSocketNIO.doIO(ClientCnxnSocketNIO.java:68)
	at org.apache.zookeeper.ClientCnxnSocketNIO.doTransport(ClientCnxnSocketNIO.java:366)
	at org.apache.zookeeper.ClientCnxn$SendThread.run(ClientCnxn.java:1145)
2019-01-28 14:33:23,675 INFO Opening socket connection to server localhost/0:0:0:0:0:0:0:1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:23,676 INFO Socket error occurred: localhost/0:0:0:0:0:0:0:1:2181: Connection refused (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:25,047 INFO [ZooKeeperClient] Closing. (kafka.zookeeper.ZooKeeperClient) [main]
2019-01-28 14:33:25,111 INFO Opening socket connection to server localhost/127.0.0.1:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:25,112 INFO Socket connection established to localhost/127.0.0.1:2181, initiating session (org.apache.zookeeper.ClientCnxn) [main-SendThread(localhost:2181)]
2019-01-28 14:33:25,218 INFO Session: 0x0 closed (org.apache.zookeeper.ZooKeeper) [main]
2019-01-28 14:33:25,218 INFO EventThread shut down for session: 0x0 (org.apache.zookeeper.ClientCnxn) [main-EventThread]
2019-01-28 14:33:25,221 INFO [ZooKeeperClient] Closed. (kafka.zookeeper.ZooKeeperClient) [main]
2019-01-28 14:33:25,223 ERROR Fatal error during KafkaServer startup. Prepare to shutdown (kafka.server.KafkaServer) [main]
kafka.zookeeper.ZooKeeperClientTimeoutException: Timed out waiting for connection while in state: CONNECTING
	at kafka.zookeeper.ZooKeeperClient.$anonfun$waitUntilConnected$3(ZooKeeperClient.scala:230)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at kafka.utils.CoreUtils$.inLock(CoreUtils.scala:251)
	at kafka.zookeeper.ZooKeeperClient.waitUntilConnected(ZooKeeperClient.scala:226)
	at kafka.zookeeper.ZooKeeperClient.<init>(ZooKeeperClient.scala:95)
	at kafka.zk.KafkaZkClient$.apply(KafkaZkClient.scala:1581)
	at kafka.server.KafkaServer.createZkClient$1(KafkaServer.scala:348)
	at kafka.server.KafkaServer.initZkClient(KafkaServer.scala:372)
	at kafka.server.KafkaServer.startup(KafkaServer.scala:202)
	at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:38)
	at kafka.Kafka$.main(Kafka.scala:75)
	at kafka.Kafka.main(Kafka.scala)
2019-01-28 14:33:25,223 INFO shutting down (kafka.server.KafkaServer) [main]
2019-01-28 14:33:25,225 WARN  (kafka.utils.CoreUtils$) [main]
java.lang.NullPointerException
	at kafka.server.KafkaServer.$anonfun$shutdown$6(KafkaServer.scala:579)
	at kafka.utils.CoreUtils$.swallow(CoreUtils.scala:86)
	at kafka.server.KafkaServer.shutdown(KafkaServer.scala:579)
	at kafka.server.KafkaServer.startup(KafkaServer.scala:329)
	at kafka.server.KafkaServerStartable.startup(KafkaServerStartable.scala:38)
	at kafka.Kafka$.main(Kafka.scala:75)
	at kafka.Kafka.main(Kafka.scala)
2019-01-28 14:33:25,231 INFO shut down completed (kafka.server.KafkaServer) [main]
2019-01-28 14:33:25,231 ERROR Exiting Kafka. (kafka.server.KafkaServerStartable) [main]
2019-01-28 14:33:25,233 INFO shutting down (kafka.server.KafkaServer) [kafka-shutdown-hook]
```

**Comment by tombentley** [comment]
It's normal for Kafka to try connecting to ZK on localhost. In order to secure communication between ZK nodes and between Kafka and ZK we use `stunnel` running in the `tls-sidecar` container to provide a secure tunnel. So the first step to debugging is checking the zookeeper logs in the zookeeper pods (are they able to elect a leader), and then checking the logs in the `tls-sidecar` containers to determine whether the problem lies there.

**Comment by scholzj** [comment]
Is this a permanent issue? Or just a one off? It happens to me quite often that one of the container images pulls slower then the first one. If it is the TLS one, it means Kafka will be already running while TLS sidecar is still pulling the image and this kind of error happens. Then after one or two restarts of the Kafka container the TLS image is finally pulled and Kafka starts well.

**Comment by BenCoughlan15** [comment]
blah, this is fine, I've got a handle on it. 

Basically, we currently have running clusters, With hundreds of clients in place, connecting via tls, etc... I was trying to get those clients to point to strimzi, without replacing/refreshing every cert. 

The problem is, another Kafka enterprise company has for a long time, recommended the use of self signed (non chained) certs,which is what we followed,  so we're in a bit of a bind. 

The hijack I'm going to use for now will be to drop the verify level from 2 to 1 on the stunnel sidecars, rest assured that I feel suitably sick for doing this, but I guess, with name-spacing permissions set properly, and ACL's enabled already, then well, I guess it's as safe as the average database. 

Long term, the solution will be to pop in a chained cert, and roll the clients properly, probably with an init container to issue a client cert every time a client comes to life. 

I'd like to do al of that with a vault PKI however, so that might be tomorrow's issue. 

Thanks for the help above.
